### PR TITLE
[5.3] Add support for field groups when rendering custom fields in the frontend

### PR DESCRIPTION
**Pull request by obuisard**
Pull Request for Issues #36967 and #36973.

### Summary of Changes

This PR addresses:
- layouts need to support the grouping of custom fields when rendered in the frontend,
- the contact field layout is missing handling of key parameters,
- the contact custom fields 'field' layout breaks the HTML markup when used inside subforms.

### Testing Instructions

In the admin console:

- [ ] Create a few field groups for articles.
- [ ] Create custom fields for articles. Put some of the fields in the field groups you have created.
- [ ] Create an article. Set values for the fields that have been created.
- [ ] Create a menu item 'Articles' -> 'Single Article'.

In the frontend:

- [ ] Check the article page before and after the patch.

Repeat with the Contacts component.

In addition, create a subform custom field for the Contacts component.
Include subform-only custom fields into the subform.
Test the output in the frontend for the subform.  The default layout will break the HTML markup.
Go back to each subform-only custom field and set the layout to 'subform' (in the 'option' tab).
The subform should be properly displayed (no more broken dl/dd/dt markup).

### Actual result BEFORE applying this Pull Request

Fields render as:

```
<ul class="fields-container">
    <li class="field-entry"> … </li>
    <li class="field-entry"> … </li>
</ul>
```

For com_contact, fields render as:

```
<dl class="fields-container contact-fields dl-horizontal">
    <dt class="contact-field-entry "> ... </dt>
    <dd class="contact-field-entry "> ... </dd>
    <dt class="contact-field-entry "> ... </dt>
    <dd class="contact-field-entry "> ... </dd>
    <dt class="contact-field-entry "> ... </dt>
    <dd class="contact-field-entry "> ... </dd>
    <dt class="contact-field-entry "> ... </dt>
    <dd class="contact-field-entry "> ... </dd>
</dl>
```

### Expected result AFTER applying this Pull Request

Fields render as:

```
<ul class="fields-container">
    <li class="field-entry"> … </li>
    <li class="field-entry"> … </li>
    <li class="field-group group-[group id]">
        <span id="group-[group id]">[group name]</span>
        <ul aria-labelledby="group-[group id]">
            <li class="field-entry"> … </li>
            <li class="field-entry"> … </li>
            <li class="field-entry"> … </li>
        </ul>
    </li>
</ul>
```

For com_contact, fields render as:

```
<dl class="fields-container contact-fields dl-horizontal">
    <dt class="contact-field-entry [field name]"> ... </dt>
    <dd class="contact-field-entry [field name]"> ... </dd>
    <dt class="contact-field-entry [field name]"> ... </dt>
    <dd class="contact-field-entry [field name]"> ... </dd>
    <dd class="contact-field-group group-[group id]">
        <span id="group-[group id]">[group name]</span>
        <dl class="fields-container" aria-labelledby="group-[group id]">
            <dt class="contact-field-entry [field name]"> ... </dt>
            <dd class="contact-field-entry [field name]"> ... </dd>
            <dt class="contact-field-entry [field name]"> ... </dt>
            <dd class="contact-field-entry [field name]"> ... </dd>
        </dl>
    </dd>
</dl>
```

For com_contact, an additional layouts/field/subform.php is present to address issues found when contact custom fields are included inside a subform (it fixes the broken markup where dl tags are missing).

### Documentation Changes Required

Yes, if there is any place where the frontend output of custom fields has been explained.

It will be necessary for the user to select the proper layout when including a contact custom field into a subform. This implies the custom field is only used in the subform.

### CODE REVIEW

**Comment by richard67** [comment]
Does this also solve #36967 ?

**Comment by obuisard** [comment]
@richard67 yes it does.

**Comment by richard67** [comment]
@obuisard Could you fix the code style errors reported by Drone here? https://ci.joomla.org/joomla/joomla-cms/51568/1/6 Thanks in advance.

**Comment by richard67** [comment]
> Does this also solve #36967 ?

Then please mention that at the top of your PR, too, and in future please try to avoid making duplicate issues if the existing issue can be found so easily. Thanks in advance.

File: components/com_fields/layouts/fields/render.php [hunk 1]
```diff
@@ -69,7 +85,47 @@
 		continue;
 	}
 
-	$output[] = '<li class="field-entry ' . $class . '">' . $content . '</li>';
+	if (!array_key_exists($field->group_id, $groupFields))
+	{
+		$groupFields[$field->group_id] = array();
+		if (Factory::getLanguage()->hasKey($field->group_title)) {
+			$groupTitles[$field->group_id] = Text::_($field->group_title);
+		}
```
Comment (richard67) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
		}
```
Remove space at end of line to fix code style error in Drone.

File: components/com_fields/layouts/fields/render.php [hunk 2]
```diff
@@ -69,7 +85,47 @@
...
+		else
+		{
+			$groupTitles[$field->group_id] = $field->group_title;
+		}
+	}
+
+	$groupFields[$field->group_id][] = '<li class="field-entry ' . $class . '">' . $content . '</li>';
+}
+
+// loop trough the groups
```
Comment (brianteeman) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
// loop through the groups
```

**Comment by brianteeman** [comment]
Shouldnt we at least try and be consistent and not have a mix of short and long array

File: components/com_fields/layouts/fields/render.php [hunk 3]
```diff
@@ -51,6 +53,16 @@
...
 
+// Organize the fields according to their group
+
+$groupFields = array(
+	0 => array()
+);
+
+$groupTitles = array(
+	0 => ''
+);
```
Comment (Quy) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
$groupFields = [
	0 => [],
];

$groupTitles = [
	0 => '',
];
```

File: components/com_fields/layouts/fields/render.php [hunk 4]
```diff
@@ -69,7 +85,48 @@
 		continue;
 	}
 
-	$output[] = '<li class="field-entry ' . $class . '">' . $content . '</li>';
+	if (!array_key_exists($field->group_id, $groupFields))
+	{
+		$groupFields[$field->group_id] = array();
+		if (Factory::getLanguage()->hasKey($field->group_title))
```
Comment (Quy) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
		$groupFields[$field->group_id] = [];

		if (Factory::getLanguage()->hasKey($field->group_title))
```

File: components/com_fields/layouts/fields/render.php [hunk 5]
```diff
@@ -59,7 +71,11 @@
 		continue;
 	}
 
-	$class = $field->name . ' ' . $field->params->get('render_class');
+	$class = $field->name;
+	if ($field->params->get('render_class'))
```
Comment (Quy) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	$class = $field->name;

	if ($field->params->get('render_class'))
```

File: components/com_fields/layouts/fields/render.php [hunk 6]
```diff
@@ -59,7 +71,11 @@
...
+	{
+		$class .= ' ' . $field->params->get('render_class');
+	}
```
Comment (richard67) [inline components/com_fields/layouts/fields/render.php]:
```suggestion
	}

```

Comment (richard67) [reply components/com_fields/layouts/fields/render.php]: Please add an empty line after the closing bracket of the control structure block.

Comment (obuisard) [reply components/com_fields/layouts/fields/render.php]: I did @richard67, thanks :-)

File: components/com_fields/layouts/fields/render.php (hunk 3 above)
Comment (obuisard) [reply components/com_fields/layouts/fields/render.php]: Thanks for the suggestions @Quy

**Comment by dgrammatiko** [comment]
There is also https://github.com/joomla/joomla-cms/pull/36265 for the grouping

**Comment by HLeithner** [comment]
This pull request has automatically rebased to 4.2-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.0-dev. No new features will be merged into Joomla! 4.3 series. Joomla! 4.4 series is a bridge release to make migration from Joomla! 4 to 5 as smooth as possible.

**Comment by ceford** [comment]
I know this is an old PR but it looked worth testing. I was rather expecting to have a custom field label look like a label - so a dt rather than a li. This is what it looks like:

screen shot 2023-09-11 at 08 27 41

Contacts - I see no html errors with/without the patch. With the patch I see this:

screen shot 2023-09-11 at 08 20 44

But the steps to get here are rather difficult to follow. I was expecting each item to be on a separate line but the source is this:

```
<dl class="fields-container contact-fields dl-horizontal">
<dt class="contact-field-entry subform-test">
            <span class="field-label ">Subform Test: </span>
    </dt>
<dd class="contact-field-entry subform-test">
        <span class="field-value ">    
    <ul class="fields-container">
        <li><span class="field-entry"><span class="field-label ">Nickname: </span>
<span class="field-value ">    Tommy</span></span>, <span class="field-entry"><span class="field-label ">Sport: </span>
<span class="field-value ">    Rugby</span></span></li>    </ul>
</span>
    </dd>
    </dl>
```
So it works but could be better!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36974">issues.joomla.org/tracker/joomla-cms/36974</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.1-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.3-dev.

**Comment by AndySDH** [comment]
Why was this closed? This was a useful improvement.

**Pull request closed without merging**